// @ilefa/husky/dist/src/util/building.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuildingCode = void 0;
var BuildingCode;
(function (BuildingCode) {
    BuildingCode["ABL"] = "Agricultural Biotech Labs";
    BuildingCode["ACS"] = "Art Ceramic Studio";
    BuildingCode["ACD"] = "Academic Building";
    BuildingCode["ACDS"] = "Academic Building";
    BuildingCode["ADC"] = "Art Design Building";
    BuildingCode["AES"] = "Architectural and Engineering Services";
    BuildingCode["APS"] = "Art Printshop";
    BuildingCode["ARJ"] = "Arjona";
    BuildingCode["ARTB"] = "Art Building";
    BuildingCode["ATL"] = "Advanced Tech Lab";
    BuildingCode["ATWR"] = "Atwater Lab";
    BuildingCode["AUST"] = "Austin";
    BuildingCode["AVPT"] = "Avery Point";
    BuildingCode["B1"] = "Biobehavioral Science #1";
    BuildingCode["B3"] = "Biobehavioral Science #3";
    BuildingCode["B4_A"] = "Biobehavioral Science #4 and Annex";
    BuildingCode["B5"] = "Biobehavioral Science #5";
    BuildingCode["BCH"] = "Beach Hall";
    BuildingCode["BRAN"] = "Branford House";
    BuildingCode["BISH"] = "Bishop";
    BuildingCode["BOUS"] = "Bousfield";
    BuildingCode["BPB"] = "BioPhysics";
    BuildingCode["BRON"] = "Bronwell";
    BuildingCode["BUSN"] = "Business";
    BuildingCode["C2E2"] = "Ctr for Clean Energy Engr";
    BuildingCode["CAST"] = "Castleman";
    BuildingCode["CHEM"] = "Chemistry";
    BuildingCode["CISS"] = "Center for Intl Students";
    BuildingCode["CPB"] = "Community & Professional Building";
    BuildingCode["CRU"] = "Cattle Resource Center";
    BuildingCode["DODD"] = "Dodd Center";
    BuildingCode["DRMU"] = "Drama/Music Building";
    BuildingCode["DWTN"] = "Stamford Downtown";
    BuildingCode["E2"] = "Engineering II";
    BuildingCode["ESB"] = "Engineering Science Building";
    BuildingCode["FG"] = "Floriculture Greenhouse";
    BuildingCode["FSB"] = "Family Studies Building";
    BuildingCode["GAMP"] = "Gampel Pavilion";
    BuildingCode["GANT"] = "Gant Central Building";
    BuildingCode["GC"] = "Gant Central Building";
    BuildingCode["GENT"] = "Gentry";
    BuildingCode["GN"] = "Gant North";
    BuildingCode["GP"] = "Gant Plaza Building";
    BuildingCode["GRE"] = "Greer Field House";
    BuildingCode["GS"] = "Gant South";
    BuildingCode["GUL"] = "Gulley Hall";
    BuildingCode["GW"] = "Gant West";
    BuildingCode["HALL"] = "Hall Dorm";
    BuildingCode["HAWL"] = "Hawley Armory";
    BuildingCode["HBL"] = "Homer Babbidge Library";
    BuildingCode["HDC"] = "Human Development Center";
    BuildingCode["HEW"] = "Hewitt";
    BuildingCode["HH"] = "Honors House";
    BuildingCode["HJT"] = "Jorgensen Theatre";
    BuildingCode["HPL"] = "Hartford Public Library";
    BuildingCode["HSSW"] = "Hartford Social Work";
    BuildingCode["HTB"] = "Hartford Times Building";
    BuildingCode["HU1"] = "Horse Unit 1";
    BuildingCode["HU2"] = "Horse Unit 2";
    BuildingCode["IMS"] = "Gant North";
    BuildingCode["IPB"] = "Innovation Partnership Building";
    BuildingCode["ITE"] = "Information Technology Engineering";
    BuildingCode["JONS"] = "Jones";
    BuildingCode["JRB"] = "J. Ray Ryan Building";
    BuildingCode["KEL"] = "Kellogg Dairy Center";
    BuildingCode["KLIN"] = "Klinck";
    BuildingCode["KNS"] = "Koons Hall";
    BuildingCode["LAFA"] = "Lafayette";
    BuildingCode["LH"] = "McHugh Hall";
    BuildingCode["LSA"] = "Life Science Annex";
    BuildingCode["LOR"] = "Lorentzon Stables";
    BuildingCode["LU1"] = "Livestock Unit 1";
    BuildingCode["LU2"] = "Livestock Unit 2";
    BuildingCode["LWB"] = "Lowell Wicker Building";
    BuildingCode["MAN"] = "Manchester Hall";
    BuildingCode["MCHU"] = "McHugh Hall";
    BuildingCode["MONT"] = "Monteith";
    BuildingCode["MSB"] = "Gant South";
    BuildingCode["MUSB"] = "Music Building";
    BuildingCode["MLIB"] = "Music Library";
    BuildingCode["OAK"] = "Herbst Hall";
    BuildingCode["PB"] = "Gant West";
    BuildingCode["PBB"] = "Pharmacy/Biology Building";
    BuildingCode["PCSB"] = "D.C. Phillips";
    BuildingCode["PR"] = "Putnam Refectory";
    BuildingCode["PU1"] = "Poultry Farm Unit 1";
    BuildingCode["PWE"] = "Pratt & Whitney Engr Building";
    BuildingCode["RHBA"] = "Ratcliffe Hicks";
    BuildingCode["ROWE"] = "Rowe";
    BuildingCode["SCI1"] = "Science One";
    BuildingCode["SCHN"] = "Schenker";
    BuildingCode["SHA"] = "Storrs Hall Annex";
    BuildingCode["SHH"] = "Herbst Hall";
    BuildingCode["SPRH"] = "Shippee Hall";
    BuildingCode["SRH"] = "Sprague Hall";
    BuildingCode["STRS"] = "Storrs Hall";
    BuildingCode["STRSWW"] = "Widmer Wing";
    BuildingCode["SU"] = "Student Union";
    BuildingCode["TAB"] = "Temporary Admin Building";
    BuildingCode["TLS"] = "Torrey Life Sciences";
    BuildingCode["TSK"] = "Tasker Building";
    BuildingCode["USRH"] = "Stamford Residence Hall";
    BuildingCode["UTEB"] = "Pratt & Whitney Engr Building";
    BuildingCode["VARC"] = "Visual Arts Resource Center";
    BuildingCode["VDM"] = "J. Louis von der Mehden Recital Hall";
    BuildingCode["WCB"] = "Wilbur Cross Building";
    BuildingCode["WGC"] = "Graduate Center";
    BuildingCode["WITE"] = "George C. White Building";
    BuildingCode["WH"] = "Wood Hall";
    BuildingCode["WREC"] = "Waterbury Rectory Building";
    BuildingCode["WSH"] = "Williams Student Health";
    BuildingCode["WSRH"] = "Wilson South Residence Hall";
    BuildingCode["WTBY"] = "Waterbury Downtown";
    BuildingCode["YNG"] = "Young";
})(BuildingCode = exports.BuildingCode || (exports.BuildingCode = {}));
